File: content-solutions/implementation/sim-cleaning-robot.py
def solution(room, r, c, d):
    rows, cols = len(room), len(room[0])
    dr = [-1, 0, 1, 0]
    dc = [0, 1, 0, -1]
    cleaned = [[False] * cols for _ in range(rows)]
    count = 0
    while True:
        if not cleaned[r][c]:
            cleaned[r][c] = True
            count += 1
        dirty = any(room[r + dr[k]][c + dc[k]] == "." and not cleaned[r + dr[k]][c + dc[k]] for k in range(4))
        if not dirty:
            br, bc = r - dr[d], c - dc[d]
            if room[br][bc] == "#":
                break
            r, c = br, bc
        else:
            d = (d + 3) % 4
            fr, fc = r + dr[d], c + dc[d]
            if room[fr][fc] == "." and not cleaned[fr][fc]:
                r, c = fr, fc
    return count

File: content-solutions/implementation/sim-cleaning-robot.js
function solution(room, r, c, d) {
  const rows = room.length;
  const cols = room[0].length;
  const dr = [-1, 0, 1, 0];
  const dc = [0, 1, 0, -1];
  const cleaned = Array.from({ length: rows }, () => new Array(cols).fill(false));
  let count = 0;
  for (;;) {
    if (!cleaned[r][c]) {
      cleaned[r][c] = true;
      count++;
    }
    let dirty = false;
    for (let k = 0; k < 4; k++) {
      const nr = r + dr[k];
      const nc = c + dc[k];
      if (room[nr][nc] === "." && !cleaned[nr][nc]) dirty = true;
    }
    if (!dirty) {
      const br = r - dr[d];
      const bc = c - dc[d];
      if (room[br][bc] === "#") break;
      [r, c] = [br, bc];
    } else {
      d = (d + 3) % 4;
      const fr = r + dr[d];
      const fc = c + dc[d];
      if (room[fr][fc] === "." && !cleaned[fr][fc]) [r, c] = [fr, fc];
    }
  }
  return count;
}
